tsfc/coffee_mode.py

```diff
@@ -0,0 +1,317 @@
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault((sum_indices, oa), []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(new_monomial_sum) == len(monomial_sum)
+
+    for (sum_indices, oa), monomials in iteritems(factor_group):
+        if len(monomials) == 1:
+            # Just one monomial with this group, add to new MonomialSum straightaway
+            monomial, = monomials
+            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+            continue
+        all_atomics = []  # collect all atomics from monomials
+        all_rest = []  # collect all rest from monomials
+        for monomial in monomials:
+            _atomics = list(monomial.atomics)
+            _atomics.remove(oa)  # remove common factor
+            all_atomics.append(_atomics)
+            all_rest.append(monomial.rest)
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            sub_monomial_sum.add((), _atomics, _rest)
+        sub_monomial_sum = optimise_monomial_sum(sub_monomial_sum, argument_indices)
+        assert len(sub_monomial_sum) > 0
+        if len(sub_monomial_sum) == 1:
+            # result is a product, add to new MonomialSum directly
+            sub_monomial, = sub_monomial_sum
+            new_atomics = sub_monomial.atomics
+            new_atomics += (oa,)  # add back common factor
+            new_rest = sub_monomial.rest
+        else:
+            # result is a sum, we need to create new node
+            new_node = monomial_sum_to_expression(sub_monomial_sum)
+            new_atomics = [oa]
+            new_rest = one
+            if set(argument_indices) & set(new_node.free_indices):
+                new_atomics.append(new_node)
+            else:
+                new_rest = new_node
+        new_monomial_sum.add(sum_indices, new_atomics, new_rest)
+    return new_monomial_sum
+
+
+def optimise_monomial_sum(monomial_sum, argument_indices):
+    """Choose optimal common atomic subexpressions and factorise a
+    :class:`MonomialSum` object.
+
+    :arg monomial_sum: a :class:`MonomialSum` object
+    :arg argument_indices: tuple of argument indices
+
+    :returns: factorised `MonomialSum` object
+    """
+    all_optimal_atomics = []  # [(sum_indces, optimal_atomics)]
+    for sum_indices in unique_sum_indices(monomial_sum):
+        # throw away other atomics here
+        optimal_atomics, _ = find_optimal_atomics(monomial_sum, frozenset(sum_indices), argument_indices)
```

Comment:
Removed. They were used in a previous version, but shouldn't need them any more in the future.